### DISCUSSION TITLE
Type liste

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#392] [General]
On a réussi à aller assez loin sans introduire de type liste, mais cette limitation du langage se fait sentir dans certains cas d'usages:

- Avoir une liste d'objets `{ HT, taux_TVA }` et calculer pour chaque item `TTC = HT × (1 + TVA)`, puis faire la somme des items `Total HT`, `Total TTC`. (retour utilisateur betagouv/publicodes#22)
- Objet `{ impactRepas, nombreDeRepas }` et calculer l'impact total de tous les repas (fonction `SUMPRODUCT` dans un tableur) (nosgestesclimat betagouv/publicodes#25). Avoir une UI spécifique pour la saisie.
- Revenu des différentes personnes d'un foyer fiscal
- Avoir un mécanisme `ajoute` à une liste pour gérer les primes conventionnelles ou les primes spécifique à un contrat https://github.com/betagouv/mon-entreprise/issues/1139#issuecomment-706527410
- Pouvoir partitionner une liste, par exemple traitement fiscal différents pour les 5 premières manifestation d'un arbitre par rapport aux suivantes https://github.com/betagouv/mon-entreprise/blob/3af23e496fd89040e3906fc10bb8d2344d537ba6/mon-entreprise/source/rules/conventions-collectives/sport.yaml#L217-L223
- Ré-écrire les mécanismes `barème`/`grile`/etc comme des mécanismes opérant sur des listes. Doit permettre de mieux composer les mécanismes au sein des tranches, par exemple rendre une tranche à taux réduit non applicable dans certains cas (utile pour l'impôt sur les sociétés par exemple betagouv/mon-entreprise#1230)
- Pouvoir calculer des résultats intermédiaire, par exemple, [un échéancier d'emprûnt](https://github.com/publicodes/publicodes/discussions/531)

On doit maintenant pouvoir imaginer la syntaxe et la sémantique d'un type liste qui améliore l'implémentation de tous ces cas d'usages.

**Comment by denismerigoux** [comment]
Salut, content de voir que vous avez finalement décidé d'ajouter ces nouvelles fonctionnalités qui devraient faciliter l'écriture de programmes publicodes. 

## Pour les objets

Ce que vous appelez "objets" désigne une structure de données avec plusieurs champs qui contiennent eux-mêmes des données. En langage théorique, c'est un "type produit" qui se décline en plusieurs formes : structures avec champs nommés, tuples. Toutes ces formes sont équivalentes (isomorphiques) d'un point de vue de la vérification de types.

Ça ne devrait pas être très dur d'ajouter ça à votre langage, c'est quelque chose de très classique. En plus ce sera complémentaire avec vos types somme.

## Le type liste

Alors là on part sur le niveau de complexité théorique du dessus. En effet, le type liste est *polymorphique* ou *générique*, car c'est toujours une liste _de quelque chose_. Donc en fait votre type liste c'est `List<T>` où `T` est un paramètre de type. Donc ça veut dire qu'il faut que votre système de types gère le polymorphisme avec ce qu'on appelle des variables de types, introduites par des des abstractions de type. On note ça avec un grand lambda classiquement: `Λ T. List<T>` qui se lit "pour un type `T` donné, le type des listes de `T`". Il va falloir que vous annotiez votre AST avec ces abstractions de types et leur applications de type correspondantes pour que votre algorithme de vérification de type marche. Vous pouvez aussi ne pas mettre ces annotations supplémentaires mais là ça revient à faire de l'inférence d'abstraction de types et l'algorithme de vérification de types devient encore plus compliqué.

**Comment by mquandalle** [comment]
Je découvre que les fonctions d'un tableur peuvent retourner une liste de valeurs, par exemple si on sélectionne 3 cellules dans la même colonne et qu'on utilise `={1;2;3}` puis <kbd>Ctrl</kbd>+<kbd>Maj</kbd>+<kbd>Enter</kbd> ça affiche les 3 nombres dans les 3 cellules. https://help.libreoffice.org/Calc/Array_Functions

Bon c'est sûrement une fonctionnalité basique des tableurs que tout le monde utilise, mais je ne connaissais pas :)

**Comment by denismerigoux** [comment]
Excel a une fonction LAMBDA maintenant pour laisser l'utilisateur définir ses propres fonctions aussi, c'est pratique :)

https://support.microsoft.com/fr-fr/office/fonction-lambda-bd212d27-1cd1-4321-a34a-ccbf254b8b67

**Comment by lajarre** [comment]
> Ré-écrire les mécanismes barème/grile/etc comme des mécanismes opérant sur des listes. Doit permettre de mieux composer les mécanismes au sein des tranches, par exemple rendre une tranche à taux réduit non applicable dans certains cas (utile pour l'impôt sur les sociétés par exemple 

Pour compléter ce que tu dis là pour barème et autres mécanismes nécessitant des listes, le cas de figure le plus simple qui ne marche pas aujourd'hui c'est de variabiliser le contenu. J'aimerais par ex pouvoir faire:

```yaml
barème:
  assiette: assiette  # no problemo
  tranches: tranches  # aïe
```

Il me semble que le fait de ne pas pouvoir le faire nous pousse régulièrement à faire des `recalcul`, ce dont il ne faut pas abuser car c'est une fonctionnalité encore très limitée et difficilement lisible (pour le dev et pour l'utilisateur final qui parcours la doc).

**Comment by johangirod** [comment]
Proposition de nouvelle fonctionnalité du langage qui résoudrait le cas d'usage proposé ici : https://github.com/publicodes/publicodes/discussions/566

Preneur de vos retours : est-ce que la solution proposée vous parait facile à comprendre ? À écrire ?
